File: src/main/java/com/vimeo/networking/model/Channel.java
package com.vimeo.networking.model;

import java.io.Serializable;
import java.util.Date;

/**
 * Created by zetterstromk on 6/11/15.
 */
public class Channel implements Serializable {

    private static final long serialVersionUID = 3190410523525111858L;

    public String uri;
    public String name;
    public String description;
    public String link;
    public Date createdTime;
    public Date modifiedTime;
    public User user;
    public PictureCollection pictures;
    public PictureCollection header;
    public Privacy privacy;
    public Metadata metadata;

    public boolean canFollow() {
        if (metadata != null && metadata.interactions != null && metadata.interactions.follow != null) {
            return true;
        }
        return false;
    }

    public boolean isFollowing() {
        if (metadata != null && metadata.interactions != null && metadata.interactions.follow != null) {
            return metadata.interactions.follow.added;
        }
        return false;
    }

    public int userCount() {
        if (metadata != null && metadata.connections != null && metadata.connections.users != null) {
            return metadata.connections.users.total;
        }
        return 0;
    }

    public int videoCount() {
        if (metadata != null && metadata.connections != null && metadata.connections.videos != null) {
            return metadata.connections.videos.total;
        }
        return 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        Channel that = (Channel) o;

        return ((this.uri != null && that.uri != null) ? this.uri.equals(that.uri) : false);
    }

    @Override
    public int hashCode() {
        return this.uri != null ? this.uri.hashCode() : 0;
    }
}

File: models/src/main/java/com/vimeo/networking2/Channel.kt
package com.vimeo.networking2

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import com.vimeo.networking2.common.Entity
import com.vimeo.networking2.common.Followable
import java.util.Date

/**
 * Channel information.
 *
 * @param categories The categories to which this channel belongs as specified by the channel moderators.
 * @param createdTime The time in ISO 8601 format when the channel was created.
 * @param description A brief explanation of the channel's content.
 * @param header The banner that appears by default at the top of the channel page.
 * @param link The URL to access the channel in a browser.
 * @param modifiedTime The time in ISO 8601 format when the album was last modified.
 * @param name The display name that identifies the channel.
 * @param pictures The active image for the channel; defaults to the thumbnail of the last video added to the channel.
 * @param privacy The privacy settings of the channel.
 * @param resourceKey The channel resource key.
 * @param tags An array of all tags assigned to this channel.
 * @param uri The unique identifier to access the channel resource.
 * @param user The Vimeo user who owns the channel.
 */
@JsonClass(generateAdapter = true)
data class Channel(

    @Json(name = "categories")
    val categories: List<Category>? = null,

    @Json(name = "created_time")
    val createdTime: Date? = null,

    @Json(name = "description")
    val description: String? = null,

    @Json(name = "header")
    val header: PictureCollection? = null,

    @Json(name = "link")
    val link: String? = null,

    @Json(name = "metadata")
    override val metadata: Metadata<ChannelConnections, ChannelInteractions>? = null,

    @Json(name = "modified_time")
    val modifiedTime: Date? = null,

    @Json(name = "name")
    val name: String? = null,

    @Json(name = "pictures")
    val pictures: PictureCollection? = null,

    @Json(name = "privacy")
    val privacy: Privacy? = null,

    @Json(name = "resource_key")
    val resourceKey: String? = null,

    @Json(name = "tags")
    val tags: List<Tag>? = null,

    @Json(name = "uri")
    val uri: String? = null,

    @Json(name = "user")
    val user: User? = null

) : Followable, Entity {
    override val identifier: String? = resourceKey
}
